File: ios/Classes/coins/BNB.swift
import WalletCore

class BNB: Coin  {
    init() {
        super.init(name: "BNB", coinType: .binance)
    }

    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {

        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)
        let publicKey = privateKey?.getPublicKeySecp256k1(compressed: true)
        
        let token = BinanceSendOrder.Token.with {
            $0.denom = "BNB" // BNB or BEP2 token symbol
            $0.amount = txData["amount"] as! Int64
        }

        let orderInput = BinanceSendOrder.Input.with {
            $0.address = AnyAddress(publicKey: publicKey!, coin: .binance).data
            $0.coins = [token]
        }

        let orderOutput = BinanceSendOrder.Output.with {
            $0.address = AnyAddress(string: txData["toAddress"] as! String, coin: .binance)!.data 
            $0.coins = [token]
        }

        let input = BinanceSigningInput.with {
            $0.chainID = txData["chainID"] as! String // Chain id (network id)
            $0.accountNumber = txData["accountNumber"] as! Int64    // On chain account / address number
            $0.sequence = txData["sequence"] as! Int64  // Sequence number, plus 1 for new order
            $0.source = txData["source"] as! Int64  // BEP10 source id
            $0.privateKey = privateKey!.data
           if (txData["memo"] != nil) {
                $0.memo = txData["memo"] as! String
            }   
            $0.sendOrder = BinanceSendOrder.with {
                $0.inputs = [orderInput]
                $0.outputs = [orderOutput]
            }
        }

        let output: BinanceSigningOutput = AnySigner.sign(input: input, coin: .binance)
        return output.encoded.hexString
    }

}

File: android/src/main/kotlin/africa/ejara/trustdart/coins/BNB.kt
import africa.ejara.trustdart.Coin
import com.google.protobuf.ByteString
import wallet.core.jni.*
import wallet.core.jni.CoinType
import wallet.core.java.AnySigner
import wallet.core.jni.proto.Binance
import africa.ejara.trustdart.Numeric
import africa.ejara.trustdart.utils.toLong


class BNB : Coin("BNB", CoinType.BINANCE) {

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        passphrase: String
    ): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        val privateKey = wallet.getKey(coinType, path)
        val publicKey = txData["fromAddress"] as String
        val signingInput = Binance.SigningInput.newBuilder()
        signingInput.chainId = txData["chainID"] as String
        signingInput.accountNumber = txData["accountNumber"]!!.toLong()
        signingInput.sequence = txData["sequence"]!!.toLong()
        signingInput.source = txData["source"]!!.toLong()
        if (txData["memo"] != null) {
            signingInput.memo = txData["memo"] as String
        }
        signingInput.privateKey = ByteString.copyFrom(privateKey.data())

        val token = Binance.SendOrder.Token.newBuilder()
        token.denom = name
        token.amount = txData["amount"]!!.toLong()

        val input = Binance.SendOrder.Input.newBuilder()
        input.address = ByteString.copyFrom(AnyAddress(publicKey, coinType).data())
        input.addAllCoins(listOf(token.build()))

        val output = Binance.SendOrder.Output.newBuilder()
        output.address =
            ByteString.copyFrom(AnyAddress(txData["toAddress"] as String, coinType).data())
        output.addAllCoins(listOf(token.build()))

        val sendOrder = Binance.SendOrder.newBuilder()
        sendOrder.addAllInputs(listOf(input.build()))
        sendOrder.addAllOutputs(listOf(output.build()))

        signingInput.sendOrder = sendOrder.build()

        val sign: Binance.SigningOutput =
            AnySigner.sign(signingInput.build(), coinType, Binance.SigningOutput.parser())
        return Numeric.toHexString(sign.encoded.toByteArray())
    }

}
